Conjur Lookup Plugin

jvanderhoof opened the pull request:
##### SUMMARY
This PR adds an Ansible Lookup Plugin which allows Ansible to retrieve secrets from Conjur using the Ansible controller's Conjur identity.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
Conjur Variable (ex. `lookup('conjur_variable', 'db/password')`

##### ANSIBLE VERSION
```
ansible 2.3.2.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 17 2016, 23:03:43) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
This lookup plugin requires Conjur (https://www.conjur.org/) to be running and accessible from the machine executing the playbook using the Conjur lookup plugin.

I wrote a small project to simplify building and validating this plugin: https://github.com/jvanderhoof/ansible-testing. It uses Docker Compose provide all the parts necessary.

jvanderhoof committed:
Imported lookup plugin from Role

jvanderhoof committed:
Plugin cleanup, including:
* Use existing Python YAML parsing
* Remove environment variables as connection options
* Added initial debugging information

jvanderhoof committed:
Reworked the lookup plugin using the Python Request library.  As it's available through Ansible, it makes communication with Conjur much more straight forward.

jvanderhoof committed:
Removed un-used libraries

jvanderhoof committed:
Fixed linting issues

Akasurde commented:
bot_broken

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+                    return yaml.safe_load(f.read())
+                except Exception as exception:
+                    display.info("Error: parsing %s - %s" % (conf_path, str(exception)))
+                    AnsibleError("Error: parsing %s - %s" % (conf_path, str(exception)))
```
Isn't `raise` missing there ? Calls to `str` are useless.

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+    def load_conf(self, conf_path):
+        conf_path = os.path.expanduser(conf_path)
+
+        if conf_path and os.path.exists(conf_path):
```
In which case first part of the condition is useful (same comment for `load_identity` method) ?

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+
+                return {"id": id, "api_key": api_key}
+            except Exception as exception:
+                AnsibleError("Error: parsing %s - %s" % (identity_path, str(exception)))
```
`raise` is missing. Call to `str` is useless there.

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+
+        return {}
+
+    def merge_dictionaries(self, *arg):
```
`merge_dictionaries` is called twice and each time with 2 dictionaries, this method could be removed and `dict.update` could be used directly.

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+
+        if identity_path and os.path.exists(identity_path):
+            display.vvvv("Loading identity from: %s" % identity_path)
+            try:
```
`try` block should contain fewer statements. `Exception` should be replaced with a more precise exception type.

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+from netrc import netrc
+from os import environ
+from time import time
+from urllib import quote_plus
```
Improve Python 3 compatibility: use [`six`](http://docs.ansible.com/ansible/devel/dev_guide/developing_python3.html#bundled-six) for `quote_plus`.

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+from urllib import quote_plus
+import yaml
+
+import requests
```
`requests` could be missing ([see](https://github.com/cyberark/ansible/blob/b304802feb6386fc362d91f694e1dce5381f8a8c/lib/ansible/plugins/callback/foreman.py#L49)). Add it to `requirements` metadata.

sivel replied:
requests should be removed completely in favor of utilizing `ansible.module_utils.urls`

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+    version_added: "2.5"
+    short_description: Fetch credentials from CyberArk Conjur.
+    description:
+      - Retrieves credentials from Conjur using the server's Conjur identity.
```
A link to server's Conjur identity website could be added.

pilou- reviewed (commented):
Could you test with Python 3 ? Have you considered adding unit tests ?

sivel commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+        display.vvvv("response: {}".format(response.text))
+
+        if response.status_code != 200:
+            raise AnsibleError('Failed to authenticate as \'{}\''.format(identity['id']))
```
This syntax is not supported in python2.6.  You will need to index your format like `{0}`

sivel commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+
+        # Use credentials to retrieve temporary authorization token
+        conjur_url = "{}/authn/{}/{}/authenticate".format(conf['appliance_url'], conf['account'], identity['id'])
+        display.vvvv("Authentication request to Conjur at: {}, with user: {}".format(conjur_url, identity['id']))
```
This syntax is not supported in python2.6. You will need to index your format like {0}

sivel commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+        )
+        if not identity:
+            raise AnsibleError('Conjur identity should be in environment variables or in one of the following paths: \'~/.netrc\', \'/etc/conjur.identity\'')
+        display.vvvv("configuration: {}".format(identity))
```
This syntax is not supported in python2.6. You will need to index your format like {0}

sivel commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,146 @@
+        headers = {'Authorization': "Token token=\"{}\"".format(token)}
+        display.vvvv("Header: {}".format(headers))
+
+        url = "{}/secrets/{}/variable/{}".format(conf['appliance_url'], conf['account'], quote_plus(terms[0]))
```
Ok, there are a bunch of these that need addressed, to index your format string for py2.6 support.

gundalow commented:
Example unit test https://github.com/ansible/ansible/tree/devel/test/units/contrib/inventory

jvanderhoof committed:
Standardized output on `format` and insure it works for 2.6, 2.7, and 3.x.

jvanderhoof committed:
Use quote_plus from the six library for improved python 2/3 behavior.

jvanderhoof committed:
Refactored identity & configuration to prefer user's file. This also includes a refactor to remove an un-needed dictionary merge method.

jvanderhoof committed:
Removed `requests` in favor of `ansible.module_utils.urls`.

jvanderhoof committed:
Refactored netrc loading to warn if host is not present.

jvanderhoof committed:
Tests and a refactor to support easier testing.

jvanderhoof committed:
Added reference to website

jvanderhoof committed:
Fixed two linting errors

jvanderhoof committed:
Fixed an extra line found by linting

mattclay commented:
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/49377/4/tests

jvanderhoof commented:
@mattclay - how can I run tests locally?  All three failures are tests I wrote.  I tested locally using python 2.7 & python 3.6 (using `virualenv`).  The way I'm running the tests is: 
```
$ pytest test/units/plugins/lookup/test_conjur_variable.py
```

mattclay commented:
@jvanderhoof Unit tests failed on both Python 3.5 and 3.6.

Run `source hacking/env-setup`, then one of the following, depending on how you want to test:

- `ansible-test units -v --docker --python 3.5 test/units/plugins/lookup/test_conjur_variable.py`
- `ansible-test units -v --tox --python 3.5 test/units/plugins/lookup/test_conjur_variable.py`

jvanderhoof committed:
Updated file write to use binary to insure config files are written correctly

jvanderhoof committed:
Resolved linting issues

jvanderhoof commented:
@sivel I've made the requested changes and added a few tests.  Mind taking another review pass?

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,148 @@
+
+# Load Conjur configuration from either `/etc/conjur.conf` or `~/.conjurrc`
+def _load_configuration():
+    for location in ['~/.conjurrc', '/etc/conjur.conf']:
```
Note that a _plugin option_ should be used for this configuration path. This would require:
- add a new parameter in plugin metadata ([example](https://github.com/ansible/ansible/blob/03794864c2c3267f1fbff6ac49e2d883090284ad/lib/ansible/plugins/connection/ssh.py#L188))
- use `get_option` method in order to fetch option value ([example](https://github.com/ansible/ansible/blob/03794864c2c3267f1fbff6ac49e2d883090284ad/lib/ansible/plugins/connection/ssh.py#L973))

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,148 @@
+
+# Load Conjur identity from either `/etc/conjur.identity`, or `~/.netrc`
+def _load_identity(appliance_url):
+    for location in ['~/.netrc', '/etc/conjur.identity']:
```
Same comment as below, a plugin option could be used there.

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,148 @@
+# (c) 2017, Jason Vanderhoof <jason.vanderhoof@cyberark.com>, Oren Ben Meir <oren.benmeir@cyberark.com>
+# (c) 2017 Ansible Project
```
It seems `# (c) 2017 Ansible Project` line should be removed (see: https://github.com/ansible/ansible/pull/34704#issuecomment-356657729).

pilou- commented:
@jvanderhoof could you add these 2 commits:
- https://github.com/pilou-/ansible/commit/0d80343b9aa2ef07c6b85aa30ec1ea15b30dfcce: updates tests to use `pytest` instead of `unittest` (all new unit tests must use [pytest](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general.html#unit-testing)),
- https://github.com/pilou-/ansible/commit/f66dffff73bf2b862905ea8c94b0cb6156401f7b: adds a mock and enables test which was commented.

jvanderhoof committed:
Refactored config & identity loading to take advantage of plugin options

jvanderhoof committed:
Cleanup a bunch of small items caught by linting

jvanderhoof committed:
Removed extra line caught by linting

jvanderhoof committed:
Swapped in pytest and added some tests with mocked network responses

mattclay commented:
CI failure in unit tests:

```
E       fixture 'mock_open_url' not found
>       available fixtures: bigip, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, mgmt_root, mock, mocker, monkeypatch, opt_bigip, opt_password, opt_peer, opt_port, opt_release, opt_token, opt_username, opt_vcmp_host, peer, pytestconfig, record_xml_property, recwarn, tmpdir, tmpdir_factory, vcmp_host, worker_id
>       use 'pytest --fixtures [testpath]' for help on them.
```

jvanderhoof commented:
@pilou, @sivel - I'm at a complete loss on how to fix the failing errors in Shippable. Here's an example:
```
                          file /root/ansible/test/units/plugins/lookup/test_conjur_variable.py, line 68
      @patch('ansible.plugins.lookup.conjur_variable.open_url')
      def test_valid_token_retrieval(self, mock_open_url):
E       fixture 'mock_open_url' not found
>       available fixtures: bigip, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, doctest_namespace, mgmt_root, mock, mocker, monkeypatch, opt_bigip, opt_password, opt_peer, opt_port, opt_release, opt_token, opt_username, opt_vcmp_host, peer, pytestconfig, record_xml_property, recwarn, tmpdir, tmpdir_factory, vcmp_host, worker_id
>       use 'pytest --fixtures [testpath]' for help on them.

/root/ansible/test/units/plugins/lookup/test_conjur_variable.py:68
```
I'm patching the `open_url` connection inside the test method, which I'm gathering is not desirable.  I can find examples of tests using @pytest.fixture, but those aren't patching methods inside the tested class (from what I can tell).

When I add a fixture method, my local tests won't run cleanly:
```py
    @pytest.fixture
    def valid_token_request(self):
        a = Mock()
        a.read.return_value = 'foo-bar-token'
        a.getcode.return_value = 200
        return a

    @patch('ansible.plugins.lookup.conjur_variable.open_url')
    def test_valid_token_retrieval(self, valid_token_request):
        response = conjur_variable._fetch_conjur_token('http://conjur', 'account', 'username', 'api_key')
        assert response == 'foo-bar-token'
```
Any help would be really appreciated.

pilou- commented:
@jvanderhoof : 
* using mock without fixture: https://github.com/pilou-/ansible/commit/0d80343b9aa2ef07c6b85aa30ec1ea15b30dfcce
* a mock based fixture:
  * [definition](https://github.com/cyberark/ansible/blob/2ca90c8a62ff32815b29b215be749bafd4cdb300/test/units/modules/packaging/os/test_rhn_register.py#L59)
  * [usage](https://github.com/cyberark/ansible/blob/2ca90c8a62ff32815b29b215be749bafd4cdb300/test/units/modules/packaging/os/test_rhn_register.py#L70)
* `pytest.mark.parametrize` should be used with `test_invalid_fetch_conjur_variable` test

jvanderhoof committed:
Pushing to see if this approach works better...

jvanderhoof committed:
Refactored be open_url mocking based on feedback

jvanderhoof committed:
Fixed a couple linting issues & refactored mocking into each method to attempt to resolve a failing test

jvanderhoof committed:
Use a generic MagicMock for python 2.6

jvanderhoof commented:
Looks like the tests ran clean with the exception of an issue retrieving a GPG key for Fedora (unrelated to any code in this PR).

Huge shoutout to @pilou for all the help with mocking.

@pilou, @sivel - mind taking another look?

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,155 @@
+      config_file:
+        description: Path to the Conjur configuration file. The configuration file is a YAML file.
+        default: /etc/conjur.conf
+        require: False
```
s/require/required/

ryanprior replied:
fixed 👍

pilou- commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,155 @@
+      _term:
+        description: Variable path
+        required: True
+      identity_file:
```
`type: path` could be added for both `identity_file` and `config_file` ([example](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/plugins/lookup/redis.py#L38)). Then : 
- `os.path.expandusers` would be called automatically (meaning calls to `expandusers` need to be removed below) 
- and `os.path.expandvars` would be called too (see [there](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/config/manager.py#L106) and [there](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/utils/path.py#L30))
- path could be relative (not sure what is the current working directory)

ryanprior replied:
I made an attempt at implementing this suggestion in https://github.com/ansible/ansible/pull/34280/commits/0fda2b504ad831e49b66fe45421006a23420838c, let me know if that's moving in the right direction.

gundalow replied:
@pilou- is https://github.com/ansible/ansible/commit/0fda2b504ad831e49b66fe45421006a23420838c correct. I've don't done any `lookup` plugin work before. I was expecting to have to see something in Python to get this working, or is the `DOCUMENTATION.foo.TYPE` process by Ansible before the lookup plugin is invoked?

pilou- replied:
It looks good.

- [read_docstring](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/plugins/loader.py#L214) is called for `lookup` plugins
- [AnsiblePlugin.get_option](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/plugins/__init__.py#L56) calls `config.get_config_value`:
    - [get_config_value](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/config/manager.py#L282) calls:
      - [get_config_value_and_origin](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/config/manager.py#L287) calls:
        - [ensure_type](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/config/manager.py#L352) calls
          - [resolve_path](https://github.com/ansible/ansible/blob/0e4e7de0001c3b53c586e6b9bafcd3f15bb862c0/lib/ansible/config/manager.py#L79)

`redis` lookup does the [same](https://github.com/ansible/ansible/blob/0fda2b504ad831e49b66fe45421006a23420838c/lib/ansible/plugins/lookup/redis.py#L40).

gundalow replied:
Thanks for educating me :)

pilou- reviewed (commented):
Some non ascii characters could be used in tests (for example: in netrc file and password).

ryanprior committed:
Fixes doc typo

require -> required

ryanprior committed:
Use `type: path` in identity_file and config_file

Also removes `expanduser` calls below (which will now be called automatically on
paths.)

ryanprior commented:
@pilou- looking forward to your further thoughts.

Also, hi! I'm a first-time Ansible contributor and excited to be involved. 👋

pilou- commented:
@gundalow @sivel unit tests have been added and comments have been addressed, could you take a look please ?

gundalow commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,155 @@
+    version_added: "2.5"
+    short_description: Fetch credentials from CyberArk Conjur.
+    description:
+      - Retrieves credentials from Conjur using the server's Conjur identity. More information about Conjur can be found at https://www.conjur.org/.
```
`U(https://www.conjur.org/)` to get a URL

ryanprior replied:
fixed

gundalow commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,155 @@
+    description:
+      - Retrieves credentials from Conjur using the server's Conjur identity. More information about Conjur can be found at https://www.conjur.org/.
+    requirements:
+      - The machine running Ansible has been given a Conjur identity.
```
I assume this process is obvious to process to people using this plugin, and therefore doesn't need further explanation?

ryanprior replied:
This will be familiar to Conjur users but I'll add a link to documentation to help people who are learning Conjur and Ansible at the same time.

gundalow reviewed (commented):
Could you also add an entry in https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml to list the maintainers.

ryanprior committed:
Defines maintainers for conjur_variable plugin

* BOTMETA.yml:
** defines $team_cyberark_conjur as maintainers of Conjur Variable plugin
** adds myself and @jvanderhoof to that team

ryanprior committed:
Adds URLs to relevant documentation for Conjur Variable lookup plugin

ryanprior commented:
@gundalow looks like shippable shows a linting error in an apparently unrelated module. Should I be concerned, is that something I can help resolve?

In any case, thanks for your comments above. Please let me know your thoughts about how I addressed them.

mattclay commented:
The unrelated errors have been fixed. Restarting CI.

jvanderhoof commented:
@mattclay - looks like there's some instability in the RPM test: 
```
 dnf-BIVXQr / /root/ansible/test/integration/targets/setup_rpm_repo/tasks/main.yml:8 / [testhost] testhost: setup_rpm_repo : Install rpmfluff and deps name={{ item }}
 failure: All items completed

{
"changed": false, 
"msg": "All items completed", 
"results": [
{
"_ansible_item_result": true, 
"_ansible_no_log": false, 
"_ansible_parsed": false, 
"attempts": 1, 
"changed": false, 
"exception": "Traceback (most recent call last):\n  File \"/tmp/ansible_IyL0Hd/ansible_module_dnf.py\", line 534, in <module>\n    main()\n  File \"/tmp/ansible_IyL0Hd/ansible_module_dnf.py\", line 530, in main\n    ensure(module, base, params['state'], params['name'], params['autoremove'])\n  File \"/tmp/ansible_IyL0Hd/ansible_module_dnf.py\", line 465, in ensure\n    base.download_packages(base.transaction.install_set)\n  File \"/usr/lib/python2.7/site-packages/dnf/base.py\", line 880, in download_packages\n    raise dnf.exceptions.DownloadError(errors.irrecoverable)\ndnf.exceptions.DownloadError: Status code: 502 for https://mirrors.fedoraproject.org/metalink?repo=updates-released-f24&arch=x86_64\n", 
"failed": true, 
"item": "python2-rpmfluff", 
"module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_IyL0Hd/ansible_module_dnf.py\", line 534, in <module>\n    main()\n  File \"/tmp/ansible_IyL0Hd/ansible_module_dnf.py\", line 530, in main\n    ensure(module, base, params['state'], params['name'], params['autoremove'])\n  File \"/tmp/ansible_IyL0Hd/ansible_module_dnf.py\", line 465, in ensure\n    base.download_packages(base.transaction.install_set)\n  File \"/usr/lib/python2.7/site-packages/dnf/base.py\", line 880, in download_packages\n    raise dnf.exceptions.DownloadError(errors.irrecoverable)\ndnf.exceptions.DownloadError: Status code: 502 for https://mirrors.fedoraproject.org/metalink?repo=updates-released-f24&arch=x86_64\n", 
"module_stdout": "", 
"msg": "MODULE FAILURE", 
"rc": 1
}
]
}
```

mattclay commented:
@jvanderhoof Yes, we do see those from time to time. I make routine sweeps of PRs failing CI to look for issues like that, so I'll make sure CI is passing, as long as the PR itself introduces no issues.

jvanderhoof commented:
@mattclay @gundalow - looks like the rest of the tests are green.

jvanderhoof commented:
@sivel - any other changes you'd like us to make in order to get this PR merged?

sivel commented:
I've approved the changes I asked to be made.  I haven't reviewed any additional changes at this time.

jvanderhoof commented:
@mattclay, @sivel, @pilou- What's the next step for this PR?  I want to make sure we can get this in for the code freeze next week.

sivel commented:
@jvanderhoof up to this point, all comments appear to largely have been code standards, guidelines, CI and such.

Generally speaking, we should get a person with Conjur experience, other than the contributor, to test and provide feedback on actual usage of this plugin.  I do not have this experience.

jvanderhoof commented:
@sivel: great point. Would this need to be someone outside CyberArk, or could it be someone internal with Ansible experience?

sivel commented:
> @sivel: great point. Would this need to be someone outside CyberArk, or could it be someone internal with Ansible experience?

I don't think it needs to be someone outside.  I'd just try to make sure you don't influence the outcome of the testing :)

jvanderhoof commented:
You don't want me to just tell them to green-light it?!?! ;)

I'm chatting internally now.  I'll keep you updated.

dustinmm80 commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,155 @@
+    lookup: conjur_variable
+    version_added: "2.5"
+    short_description: Fetch credentials from CyberArk Conjur.
+    description:
```
Can we update the `description` text to be a little more explicit that it is the controlling host (the machine running Ansible) and not the target node that must have an established Conjur identity for the lookup plugin to work?

Right now it's a bit confusing because the `requirements` is clear about that but the description just says `...the server's Conjur identity` without saying what 'the server' is in this context.

ryanprior replied:
done

garymoon commented:
I have two suggestions:

1. If we could surface any non-200 status codes at https://github.com/ansible/ansible/pull/34280/files#diff-0aef6d55c9011ffcdda37eafa3ca0e85R115 it could give users a good indication of what's causing the issue.
2. If we could restate that the lookup is running on the controlling host instead of the target node in places where there's an issue getting the Conjur identity or connecting to the server, that might head some lengthy troubleshooting off at the pass. See:
https://github.com/ansible/ansible/pull/34280/files#diff-0aef6d55c9011ffcdda37eafa3ca0e85R92
https://github.com/ansible/ansible/pull/34280/files#diff-0aef6d55c9011ffcdda37eafa3ca0e85R77
https://github.com/ansible/ansible/pull/34280/files#diff-0aef6d55c9011ffcdda37eafa3ca0e85R114
https://github.com/ansible/ansible/pull/34280/files#diff-0aef6d55c9011ffcdda37eafa3ca0e85R135 and below

ryanprior committed:
Clarifies "the server," "the machine" -> "controlling host"

The machine identity used is that of the Ansible controlling host, not any
server being provisioned or instructed. This documentation change aims to make
that relationship clear.

ryanprior committed:
Adds response code to exception message on authentication failure

ryanprior committed:
Enhances exception messages to specify the controlling host

These error messages are less likely to confuse a user as to which machine is
associated with the files, identities, and configurations being described.

ryanprior commented:
Thank you @garymoon and @dustinmm80 for helping to validate the plugin. Let me know if you have any questions as to how I addressed your feedback.

@sivel same goes for you, please let me know if this round of patches raises any concerns.

dustinmm80 commented:
These changes all look good to me, thanks Ryan.

This lookup plugin is helpful when configuring nodes with secrets when the controlling host has been assigned a Conjur identity. It uses the standard paths for Conjur identity and configuration files, that's nice.

My review is 👍 , let's ship it!

garymoon commented:
LGTM 👍

jvanderhoof commented:
@sivel - we received and resolved some feedback from two team members not involved in this effort, but with significant Ansible experience.  Anything else we can do to get this in?

pilou- commented:
@jvanderhoof in order to get rid of [`needs_revision`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#label-needs_revision) label, you could comment using [`ready_for_review`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#cmd-ready_for_review).

jvanderhoof commented:
ready_for_review

jvanderhoof commented:
#ready_for_review

jvanderhoof commented:
bot_status

sivel commented:
We've "identified" what appears to be an issue with GitHub on a few pull requests, where the UI indicates that the PR is "green" (which is correct), but "mergeable_state" is indicated to be "unstable".  Although this PR currently states "needs_revision" it should be fine.  Removing it will just cause the bot to re-add that label.

jvanderhoof commented:
@sivel - what does that mean for this PR?

sivel commented:
@jvanderhoof nothing specifically, other than I am saying if the PR is "green" we can ignore "needs_revision".

It just needs final reviews.  I will personally not have time today to get to this.

thaumos commented on lib/ansible/plugins/lookup/conjur_variable.py:
```diff
@@ -0,0 +1,163 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
```
@jvanderhoof or @ryanprior,

could you add below here?

```
ANSIBLE_METADATA = {'metadata_version': '1.1',
                    'status': ['preview'],
                    'supported_by': 'community'}
```

ryanprior replied:
Added

jvanderhoof replied:
@thaumos - Thanks for the feedback. Mind giving it one more look? I'd love for this to be able to make in today if possible with code freeze on the 22nd.

thaumos replied:
lgtm.  there shouldn't be a concern having this merged before code freeze.

ryanprior committed:
Adds ANSIBLE_METADATA for Conjur variable lookup plugin

sivel commented:
Next steps?  I'm going to click the big green button.

jvanderhoof commented:
A big thanks to @sivel, @thaumos, @mattclay, @pilou- & the rest of Ansible core team.  We really appreciate your help and feedback!
